WIP: Auto add PRs to project if they match a search query

wwwil opened the pull request:
This is in response to #9925.

Currently implemented as a plugin called `project-queries`. It's configured in `plugins.yaml`, with a global section listing projects to add to, each with a search query.

```
project_queries:
  - name: my-project
    org: my-organisation
    column: todo
    search_query:
      state: open
      labels:
        - todo 
      org: my-organisation

plugins:
  my-organisation:
  - project-queries
```

The plugin responds to pull request events. It checks whether the pull request matches any search queries, and for each match creates a card on the corresponding project.

It still needs some work:
*  Add tests
*  Allow list of orgs in search query

/kind feature

stevekuznetsov commented on prow/github/client.go:
```diff
@@ -2316,3 +2316,75 @@ func (s *reloadingTokenSource) Token() (*oauth2.Token, error) {
 	}, nil
 }
+
+// GetOrgProjects returns the list of projects in this org.
```
Could you please add the documentation link to this comment

wwwil replied:
I've added the documentation links for `GetOrgProjects` and `GetProjectColumns`.

stevekuznetsov commented on prow/github/client.go:
```diff
@@ -2316,3 +2316,75 @@ func (s *reloadingTokenSource) Token() (*oauth2.Token, error) {
+		return nil, errors.New("projectCard.ContentType must be either Issue or PullRequest")
+	}
+
+	// Not every create func has this block, TODO find out why and whether it should be here
```
The `dry` should be on any call, but `fake` seems leaky -- @fejta how many places use the global fake versus making their own scoped ones?

fejta replied:
Not sure what you mean? I think the fake is mostly vestigial at this point?

wwwil replied:
I've removed the reference to `fake` but left the reference to `dry`.

stevekuznetsov commented on prow/plugins/plugins.go:
```diff
@@ -221,6 +221,7 @@ type Configuration struct {
 	Size                 *Size                  `json:"size,omitempty"`
 	Triggers             []Trigger              `json:"triggers,omitempty"`
 	Welcome              []Welcome              `json:"welcome,omitempty"`
+	ProjectQueries       []ProjectQuery         `json:"project_queries,omitempty"`
```
I would not necessarily call these `queries` as those have a specific technical definition when in the context of the GitHub API

stevekuznetsov replied:
Instead of having a loose list of queries, let's key these either on org or org/repo -- look at the BranchProtection configuration. That sort of tree structure with overrides is more natural for configuring this sort of behavior and from our experience easier for users to interface with.

wwwil replied:
Would that be keyed to the org/repo of the project? So for example, to add to the KEP Tracking project you would add the following to `plugins.yaml`:
https://github.com/kubernetes-sigs/architecture-tracking/projects/2
```
project_queries:
  orgs:
    kubernetes-sigs:
      repos:
        architecture-tracking:
          projects:
            KEP Tracking:
              columns:
                Backlog:
                  state: open
                  org: kubernetes
                  labels:
                    - kind/kep
```
Or possibly something a bit less verbose, but along those lines?

stevekuznetsov replied:
Yeah, something like that. If there's no sense in having projects that span multiple repos we can just key on `org/repo`

stevekuznetsov replied:
I'd also suggest something like:

```yaml
projects:
- name: "KEP Tracking"
```

Instead of 

```yaml
projects:
  KEP Tracking:
```

For ease of use once it's ingested into Go structs

fejta replied:
On the other hand using the `projects: {"KEP Tracking": ...}` pattern makes it easier to prevent duplicate names.

wwwil replied:
I'm now thinking of structuring it like this:

```
project_queries:
  org/repo:
    kubernetes-sigs/architecture-tracking:
      projects:
      - name: "KEP Tracking"
        columns:
        - name: "Backlog"
           state: open
           org: kubernetes
           labels:
           - kind/kep
```

This groups org and repo so the yaml gets less indented and should be easier to read and recognise org/repos. Org level projects can also be referenced by not specifying a `/`.

Each org/repo then has a list of projects so that easily translates to a `map[string][]Project`. Each project has a name and list of columns, and each column has a name and set of conditions so these easily translate to a `project` and `column` struct.

As a side note I plan to change the structure of the `column` part so `org` is `orgs` and a list of orgs to add PRs from can be given. Other terms could also be added to this as required for more flexible matching.

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+// These are the only actions that this plugin cares about
+func isPRChanged(pe github.PullRequestEvent) bool {
```
I'm not sure this method brings much value over:

```go
if pe.Action != github.PullRequestActionOpened {
    return nil
}
```

At the callsite

wwwil replied:
Wouldn't this miss pull requests that had relevant labels added after opening?

stevekuznetsov replied:
I guess the larger point was -- we can enumerate the list of actions in the `if` statement as done in the other plugins, not much benefit to break out into a separate method and explicitly deny other actions

wwwil replied:
I've now changed this to use an if statement and removed the `isPRChanged` function.

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+
+func handlePR(gc githubClient, projectQueries []plugins.ProjectQuery, le *logrus.Entry, pe github.PullRequestEvent) error {
+
+	if len(projectQueries) == 0 {
```
Should config load fail when you turn on a plugin but it has no behavior because you've not configured it?

stevekuznetsov replied:
@cjwagner @fejta

cjwagner replied:
Yes, config load should fail in that case. That would never have the effect that the user wants so it should be considered a misconfiguration.

wwwil replied:
Would that mean it should error here rather than just return `nil`? Or should an error be raised elsewhere?

stevekuznetsov replied:
In `prow/config` package on the configuration loading and validating section

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+	if err != nil {
+		return err
+	}
+	if len(matchedProjectQueries) == 0 {
```
You'll get this behavior by default with the `for _, matchedProjectQuery := ` loop below if the slice is empty, so you don't need this block

wwwil replied:
I've removed the redundant block.

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+
+	for _, matchedProjectQuery := range matchedProjectQueries {
+		err = addPRToProject(gc, matchedProjectQuery, pe)
+		le.WithError(err)
```
Generally we call `*logrus.Entry` `log` or `logger`.

wwwil replied:
Renamed to `log`.

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+
+// labelsInLabels check that all of the queryLabels (the first list), are present in
+// the list of github.Labels (the second list)
+func labelsInLabels(queryLabels []string, labels []github.Label) bool {
```
You could use:

https://github.com/kubernetes/test-infra/blob/2f2f372c620afd5c000f5068d89db21f3865ddf6/prow/github/helpers.go#L26-L34

Or add your own there. If adding your own, consider using `sets.String.HasAll()`

wwwil replied:
I have added a `HasLables` function to `helpers.go`, changed the `project_queries` plugin to use the helper function and removed the redundant `labelsInLabels` function.

The new `HasLabels` function iterates through the list of strings, making use of the existing `HasLabel` function for each rather than using `sets.String.HasAll()`. This is to ensure consistent behaviour as `HasLabel` compares both strings converted to lower case.

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+}
+
+func getMatchingProjectQueries(gc githubClient, projectQueries []plugins.ProjectQuery, pe github.PullRequestEvent) ([]plugins.ProjectQuery, error) {
+	matchedProjectQueries := make([]plugins.ProjectQuery, 0, 1)
```
Generally unless we have an explicit reason to use`make` like this, a nil slice instantiation line is fine:

```go
var matchedProjectQueries []plugins.ProjectQuery
```

wwwil replied:
Done!

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+func getMatchingProjectQueries(gc githubClient, projectQueries []plugins.ProjectQuery, pe github.PullRequestEvent) ([]plugins.ProjectQuery, error) {
+	matchedProjectQueries := make([]plugins.ProjectQuery, 0, 1)
+	// GetIssueLabels works for PRs as they are considered issues in the API
+	labels, err := gc.GetIssueLabels(pe.Repo.Owner.Login, pe.Repo.Name, pe.Number)
```
This is one of the items I was asking about in the issue -- we need to be careful on how this consumes GitHub API tokens. I would suggest making this lazily resolve labels only when we are certain we need to know them to determine if a project should be added.

wwwil replied:
I've changed this now to only use `GetIssueLabels` at the latest possible point, if required to determine a match.

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+
+func addPRToProject(gc githubClient, projectQuery plugins.ProjectQuery, pe github.PullRequestEvent) error {
+	// Get the project id by name
+	projects, err := gc.GetOrgProjects(projectQuery.Org)
```
For deployments without the GitHub proxy cache, the projects and columns seem like a good place to add an in-memory cache

wwwil replied:
So would the plugin just hold onto a list of projects outside of any functions? Then it could query this first and only using `GetOrgProjects` if it can't match a project name. Or its there a particular way a cache should be implemented?

stevekuznetsov replied:
That seems reasonable -- look at the cache in `prow/commentpruner/commentpruner.Go#EventClient.comments` for an example

wwwil replied:
I've now added local cache of `github.Projects` and `github.ProjectColumn` to the plugin.

A struct called  `projectWithColumns` is used to hold a `github.Project` with a slice of `github.ProjectColumn`s.

The plugin holds a map of `orgRepo` strings to slices of `projectWithColumns`. Each time a PR is added to a project this map is searched first. If the project cannot be found the slice for that `orgRepo` string is updated from GitHub using `GetOrgProjects`, and the search is repeated. This process is repeated when searching for the project's column to add the PR to.

Additionally when the projects for an `orgRepo` are updated and a new slice of `projectWithColumns` is made, the columns for any project whose ID stays the same should be kept from the old slice, reducing the need to get columns from GitHub with `GetProjectColumns`. However this doesn't seem to be working currently as looking at logs `GetProjectColumns` is still called every time a PR is added to a project, so I'm still working on that.

I have also noticed I still need to add a way to get a list of GitHub projects at the repo level.

stevekuznetsov commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,208 @@
+	projectCard := github.ProjectCard{}
+	projectCard.ContentType = "PullRequest"
+	projectCard.ContentID = pe.PullRequest.ID
+	_, err = gc.CreateProjectCard(columnID, projectCard)
```
What happens if there exists a project definition that is owned by this plugin, and the PR used to match the criteria but no longer does? Do we want to remove it?

wwwil replied:
At the moment it just adds pull requests. Removing them or moving them along the columns has to be done manually.

If it could remove pull requests that would be nicer, and would even mean things could progress through columns automatically.

Removing them would mean either keeping track of automatically added pull requests, or removing any pull request that doesn't match the query. Removing anything that doesn't match could be annoying, as if for some reason a pull request was manually added that doesn't match the query it would be removed by the plugin.

stevekuznetsov replied:
Can we declare that for Projects managed by this plugin, only automatic PR additions are allowed? What do we need for the UX here?

stevekuznetsov replied:
@cblecker

munnerz commented:
/cc

wwwil commented:
Thanks @stevekuznetsov for the review feedback, much appreciated and very helpful!

wwwil commented on prow/hook/BUILD.bazel:
```diff
@@ -57,7 +57,11 @@ go_library(
         "//prow/plugins/milestonestatus:go_default_library",
         "//prow/plugins/override:go_default_library",
         "//prow/plugins/owners-label:go_default_library",
+<<<<<<< HEAD
```
Oops! I've now fixed this.

wwwil commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,181 @@
+			continue
+		}
+		if len(searchQuery.Labels) != 0 {
+			if !labelsFetched {
```
If there were no labels then the `labels` slice would always have a length of `0` so it would keep calling `GetIssueLabels`.

taragu commented on prow/github/client.go:
```diff
@@ -2430,3 +2430,79 @@ func (s *reloadingTokenSource) Token() (*oauth2.Token, error) {
+// GetOrgProjects returns the list of projects in this org.
+//
+// See https://developer.github.com/v3/projects/#list-organization-projects
+func (c *Client) GetOrgProjects(org string) ([]Project, error) {
```
Does `GET /orgs/:org/projects` get all projects under each repo of an organization, or just the projects on the organization level?

wwwil replied:
`GET /orgs/:org/projects` only gets the projects at the organisation level. I've added another function to use `GET /repos/:owner/:repo/projects` to get repo level projects as described here: https://developer.github.com/v3/projects/#list-repository-projects.

fejta commented on prow/github/helpers.go:
```diff
@@ -33,6 +33,16 @@ func HasLabel(label string, issueLabels []Label) bool {
 }
 
+// HasLabels checks if all labels are in the label set "issueLabels"
+func HasLabels(labels []string, issueLabels []Label) bool {
```
Only one package uses this function, please move this to that package and make it private to that package

wwwil replied:
Done!

fejta commented on prow/plugins/projectqueries/projectqueries.go:
```diff
@@ -0,0 +1,251 @@
+func addPRToProject(gc githubClient, projectQuery plugins.ProjectQuery, pe github.PullRequestEvent) error {
+	// Try to find the current project by name
+	currentProject := getProjectByName(projectQuery.Name, orgRepoProjects[projectQuery.Org])
+	// If the project could not be found in the local list then update the it
```
Why do we need the caching here? It seems like it would be simpler to just make the API call every time, rather than maintain a potentially invalid list.

We have a ghproxy to cache get requests to solve this sort of thing.

wwwil replied:
The local caching of projects was suggested by @stevekuznetsov to reduce API calls. I'm quite new to the project so not sure what the best practise is for this kind of thing.

fejta replied:
/cc @cjwagner 
thoughts?

I am always confused about when we do, do not care about tokens

cjwagner replied:
If you can be sure that cache won't ever serve stale data then definitely keep it.
We don't want to rely on `ghproxy` if we don't have to so that smaller Prow instances don't need to use it and so that when `ghproxy` is unavailable or has an empty/stale cache, Prow can still operate normally. (e.g. after switching bot accounts or tokens).

`ghproxy` greatly improves our token usage, but it is a single point of failure if Prow becomes reliant on it.

wwwil replied:
I'm now considering changing the plugin config to refer to projects only by number, and columns by ID. This removes the need to make all the API calls to convert names to IDs, or the need for a cache which could become invalid.

Project numbers can be easily found in the GitHub UI. An API call will still be needed to get the actual ID from the org/repo and project number but this can be cached and should never become invalid as you can't change a project number.

Column IDs can be found in the URL obtained using "Copy column link" in the "..." menu.

If the ability to use names was still desired I would suggest that it not use any caching, so users can choose between using numbers and IDs with lower API usage, or full names with more API usage.

fejta replied:
> We don't want to rely on ghproxy if we don't have to

IMO we want to rely on ghproxy whenever we have to, assuming it involves caching. We don't want every plugin to roll its own cache. This is a recipe for bugs and unnecessary complexity.

If we need caching please depend on ghproxy, don't create something new.

The assumption is either prow is managing repos slow enough that caching doesn't matter. Otherwise deploy ghproxy.

> changing the plugin config to refer to projects only by number

This sounds unfriendly to the people using the plugin.

> projects in an org or org/repo aren't required to have unique names

Technically teams also aren't required to be unique, but since this is hard to work with, peribolos is just going to enforce that team names are unique. I would suggest we do the same thing here (aka if someone references foo then find the first foo project in org/repo else org that matches)

cjwagner commented:
/ok-to-test

fejta commented:
Please rebase and try to get tests working (try out `hack/check-pr.sh`)

wwwil commented:
@fejta I've just rebased, made a bit of a mess as I didn't notice things had changed with `/prow/plugins/plugins.go` but I think I've now fixed that. I'm going to work on tests next, hopefully will have some time for that this weekend!

stevekuznetsov commented on prow/plugins/projectmanager/projectmanager.go:
```diff
@@ -0,0 +1,204 @@
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns if they match given criteria.",
+		Config: func(config *plugins.Configuration) map[string]string {
```
Could we achieve this with `json.Marshal`?

cblecker commented:
It's not clear to me from all the review comments what the configuration would look like here. Can we have the PR body updated with that? I want to understand what the config would look like for repo level projects, and org level projects.

fejta commented:
Do you have an ETA for the next update of this PR?

dims commented:
@wwwil ping :)

wwwil commented:
Hi everyone, sorry for my non responsiveness, recently started a new job and moved into a new flat so been very busy! I'm planning to resume work on the plugin next week.

brahmaroutu commented:
/assign srbrahma
@wwwil I will be working on this PR as a separate PR, we would like to push this code in ASAP.

brahmaroutu commented:
https://github.com/kubernetes/test-infra/pull/12326 replicates this PR.
We can close this PR and carry on work in the new PR.
